corpus/expressions.txt

```diff
@@ -1877,9 +1877,9 @@ var x = c is < '0' or >= 'A' and <= 'Z';
           (equals_value_clause
             (is_pattern_expression
               (identifier)
-              (binary_pattern
+              (or_pattern
                 (relational_pattern (character_literal))
-                (binary_pattern
+                (and_pattern
```

Comment:
What is the rationale for removing `binary_pattern`? It looks like the change doesn't change which programs are accepted, and changing it potentially breaks any existing consumers of the syntax tree. I think this change should be reverted.

Reply:
Up until this point there hasn't been a specific goal on tree-sitter-c-sharp to retain every published named rule.

If that's a goal across tree-sitter i is mostly doable but introduces extra complexity in trying to keep up with new C# grammar rules where right now we try and stick close to the Roslyn grammar (as tricky as that is to piece together given that the official spec is always many versions behind the real language).

Reply:
Thanks for clarifying! I think these things are somewhat underspecified in the tree-sitter ecosystem atm :) Alignment with some reference is definitely valuable, so if this achieves that, I think that's a valid reason to go through with it.

Reply:
Okay, dug a bit more into this (the PR was in draft for a while so I'd forgotten most of it).

The grammar such as it is documented has kept the `binary_pattern` for or and and *but* I was unable to get them to parse correctly as I needed to deal with precedence so split them so each could have their own.

In the VS Syntax Analyzer (primary way I figure out what Rolsyn is *actually* doing) I see that `OrPattern` and `AndPattern` are actually separately specified.

e.g.

```csharp
var a = p is [1, 2, x,] and [] or [2, ..];
```

Corresponds to 
![image](https://user-images.githubusercontent.com/118951/187968866-48d0efc8-f668-4d0d-b71c-5ac7e89d84a1.png)

Reply:
That makes sense. And thanks for the extensive explanation ;)